Implement Unix domain socket support for VLAN

arixmkii opened the pull request:
Fixes #15852 

This change adds support for new QEMU stream netdev added in 7.2.0. It is implemented as an opt-in mode for previously supported platforms and the only supported mode on Windows.

Old FD netdev has changes only on podman side. Instead of previously used QMP socket address, it now has VLAN dedicated socket address to make both implementations more similar.

As VLAN socket address has short lifespan (it exists only after forwarder has been started and before QEMU has finished startup), it is not promoted to persisted machine settings, but is rather calculated inside Start method.

Signed-off-by: Arthur Sengileyev <arthur.sengileyev@gmail.com>

I tried to highlight everything significant in the commit message. This indeed fixes #15852, but it has a lot more added that the change required to fix it. I needed other changes to actually check the desired behavior of the fix for the issue mentioned above, if the team will insist I can split it up into series of commits (I would prefer not to, though 😅 ).

VLAN socket address is not persisted anywhere in machine socket, so, the logic decides on the mode by checking the command line. Actually the way how QEMU command line is stored is dictating a lot of how QEMU and in turn gvproxy and Podman should operate. So, it might look dirty, but to me looked like a reasonable implementation (at least at this point in time).

**Machine Init**

When in SOCKET VLAN mode (on platforms, where FD supported it has to be enabled via env var) the Init will create QEMU command line compatible with new QEMU 7.2.0 feature.

**Machine Start**

Podman will ignore the current settings, but will rely on what was create by Machine Init (analyze command line)

**Named pipes**

Now gvproxy will also expose named pipe on Windows (no-op for non-Windows paths). The name of the pipe is of the format `qemu-podman-...` For default machine it will result in `qemu-podman-machine-default`. This prefix is needed
to prevent name collision with WSL machine. Clients could read pipe name from `podman machine inspect` output (the change is already merged to Podman Desktop).

**Mode switch**

Current implementation is using opt-in via env var, but alternative implementations (like checking QEMU version or config files) could exist. Socket VLAN is more cross platform, but it can't replace old implementation, because
it is only available starting from QEMU 7.2.0 (which means baseline RHEL 10 😅 )

**Configuration mismatch warning**

When mode switch has happened between Init and Start - like machine was created with FD settings, but then Start is
called, when SOCKET VLAN usage is requested - there will be WARN output what could be changed to fix the mismatch.
Works in both directions.

**Socket VLAN gvproxy readiness check**

It is impossible to use the Dial, because then we could not pass it forward as FD. The most common reason for failed start
would be socket address unavailable. In this case gvproxy will terminate. Readiness check algorithm looks like this:
1. Initial delay (small sleep) to account for gvproxy starting up in the background.
2. Start loop for up to max attempts
3. Check if process is alive using PID
  3.1. If not, then short circuit to failure
  3.2. If process is running continue to the next check
4. Check if we can `os.Stat` the socket address
 4.1. If not then sleep and go to the next loop iteration
 4.4. If ok, then break the look and continue
 
Refactorings
* startHostNetworking private method is now more sophisticated implementation: accepts input param, returns in addition Process data and has logic changes to account for name pipe functionality.

Additional fixes
* use "reasonable" UID on Windows (no platform check, but we rather check for the invalid return value)
* fixed type in Errorf in `checkProcessStatus` in _unix sources
* use stored `v.ReadySock.Path` instead of making it up again in Start method using the same rules

Local testing

Acceptance test: start the machine and run nginx with port forwarding to host machine.

* Tested on Windows applying all other pending changes for Windows support first
* Tested on Apple Silicon Mac

Output from Apple Silicon Mac:
**Running FD created machine (created with release version of Podman) w/ FD mode**
```
podman % ./bin/darwin/podman --log-level=debug machine start
INFO[0000] ./bin/darwin/podman filtering at log level debug
Starting machine "podman-machine-default"
[/Users/username/git/podman/bin/libexec/podman/gvproxy -listen-qemu unix:///var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/vlan_podman-machine-default.sock -pid-file /var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/podman-machine-default_proxy.pid -ssh-port 64042 -forward-sock /Users/username/.local/share/containers/podman/machine/podman-machine-default/podman.sock -forward-dest /run/user/501/podman/podman.sock -forward-user core -forward-identity /Users/username/.ssh/podman-machine-default --debug]
DEBU[0000] qemu cmd: [/opt/homebrew/bin/qemu-system-aarch64 -m 8192 -smp 4 -fw_cfg name=opt/com.coreos/config,file=/Users/username/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix:/var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=apodman-machine-default_ready -device virtserialport,chardev=apodman-machine-default_ready,name=org.fedoraproject.port.0 -pidfile /var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/podman-machine-default_vm.pid -accel hvf -accel tcg -cpu host -M virt,highmem=on -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/username/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/username,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/username/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-37.20230110.2.0-qemu.aarch64.qcow2] 
Waiting for VM ...
...
```

**Running FD created machine (created with release version of Podman) w/ socket VLAN mode**
```
podman % CONTAINERS_USE_SOCKET_VLAN=true ./bin/darwin/podman --log-level=debug machine start
INFO[0000] ./bin/darwin/podman filtering at log level debug 
Starting machine "podman-machine-default"
WARN[0000] Stored Podman Machine configuration doesn't match current settings 
WARN[0000] Consider replacing "socket,id=vlan,fd=3" with "stream,id=vlan,server=off,addr.type=unix,addr.path=/var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/vlan_podman-machine-default.sock" in machine config 
[/Users/username/git/podman/bin/libexec/podman/gvproxy -listen-qemu unix:///var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/vlan_podman-machine-default.sock -pid-file /var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/podman-machine-default_proxy.pid -ssh-port 64042 -forward-sock /Users/username/.local/share/containers/podman/machine/podman-machine-default/podman.sock -forward-dest /run/user/501/podman/podman.sock -forward-user core -forward-identity /Users/username/.ssh/podman-machine-default --debug]
DEBU[0000] qemu cmd: [/opt/homebrew/bin/qemu-system-aarch64 -m 8192 -smp 4 -fw_cfg name=opt/com.coreos/config,file=/Users/username/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix:/var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=apodman-machine-default_ready -device virtserialport,chardev=apodman-machine-default_ready,name=org.fedoraproject.port.0 -pidfile /var/folders/9n/358_gdmn3fxfnyttwys8k1dm0000gn/T/podman/podman-machine-default_vm.pid -accel hvf -accel tcg -cpu host -M virt,highmem=on -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/username/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/username,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/username/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-37.20230110.2.0-qemu.aarch64.qcow2] 
Waiting for VM ...
...
```

[NO NEW TESTS NEEDED]

<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Podman Machine now supports QEMU 7.2.0 stream netdev for VLAN socket
```

arixmkii commented:
It should not alter any default behavior and everything else is ether behind opt-in toggle or not yet supported platform (QEMU on Windows), so, I was leaning towards adding NO NEW TESTS tag, but the changeset is not small/trivial, so, the team might disagree with it.

This changeset with all other in pending review states is 99% of changes needed to enable QEMU on Windows.

But yeah, I don't expect this PR would be the fast one to get through.

arixmkii commented:
I'm not sure if this part https://github.com/containers/podman/blob/06e85f0f6093081e8707b30e191ba190613090b1/pkg/machine/qemu/machine.go#L502-L512 behaved on Windows as expected, but for sure it didn't affect happy path in my testing.

arixmkii commented:
Rebased to main to resolve conflicts from another merged PR.

arixmkii commented:
I'm still a bit lost if it is possible to write a test for this one as one needs to use Windows for it or being able to modify environment variable for a specific test.

arixmkii commented:
@rhatdan may be you can help me to find some reviewers for this one? At least to get some comments if it is ok conceptually.

rhatdan commented:
@baude @Luap99 @n1hility @ashley-cui PTAL

TomSweeneyRedHat commented on pkg/machine/qemu/machine.go:
```diff
@@ -507,29 +514,87 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
+			return err
+		}
+
+		for i := 0; i < 6; i++ {
```
why 6?  Personal nit, I'd prefer a variable with a name that explained why 6.

arixmkii replied:
Just because it was already used in this file in 2 places https://github.com/containers/podman/blob/main/pkg/machine/qemu/machine.go#L515 and https://github.com/containers/podman/blob/main/pkg/machine/qemu/machine.go#L604

Probably it should refactored into more generic utility, which could be used in all places in this file. I have not strong opinion if this should be done as part of this PR or not. If it is agreed to add the refactoring here, then I will do it in a separate commit as this is not implementation related.

arixmkii replied:
Added named constants and introduced unified retry function in a separate commit.

TomSweeneyRedHat commented:
The changes in general LGTM, but this is all way out of my wheelhouse

n1hility commented:
> I'm not sure if this part behaved on Windows as expected, but for sure it didn't affect happy path in my testing.
> 
> https://github.com/containers/podman/blob/06e85f0f6093081e8707b30e191ba190613090b1/pkg/machine/qemu/machine.go#L502-L512

This should be safe on Windows. The biggest difference is that SIGTERM can't be eaten and the program kept alive, you can only use it for cleanup or exiting with an exit code like this block is doing, so this part looks right to me.

arixmkii commented:
I don't like that I had to modify outer scope variables from the function being retried, but w/o generics it is not possible to implement configurable compile time checked return type.

rhatdan commented:
Either add tests or [NO NEW TESTS NEEDED]

arixmkii commented:
I will add  [NO NEW TESTS NEEDED] and force push. My reasoning:
* Windows QEMU machine is not enabled yet (and it will require additional setup for tests, a bit topic on its own)
* Alternate mode of FD will be purely opt in and has stricter requirements on minimal supported QEMU version, it looks like it now can only be reliably tested on macOS with brew

arixmkii commented:
Observed some sort of regression at least in my test Windows build, so, I might apply some fixes to the commits, but the concept is not expect to change.

arixmkii commented:
@TomSweeneyRedHat @n1hility are any additional changes needed?

arixmkii commented:
Failed "Verify Win Installer Build" looks like a flake to me (at least from build log and searching the web for that specific error).

n1hility commented:
Looking into the verify installer failure, we have other runs where its working and repeated runs are failing here, so looking into why that happens

arixmkii commented:
@n1hility I can do rebase. Just ping me if it is needed.

n1hility commented:
@arixmkii sure if you have a chance lets try that

n1hility commented:
FYI I needed to add a debug commit to your branch since I cant reproduce this failure in my local repo via cirrus

/hold

n1hility commented:
The win installer validation problem triggered by this PR is resolved. There is a separate PR for the fix, but I also put the commit on here in the meantime to reflect the true result of the testing.

I still need to review this in more detail. Promise to do that tomorrow..

TomSweeneyRedHat commented:
Happy green test buttons.  The changes LGTM, but I'd like more eyes on this, especially @n1hility

n1hility commented on pkg/machine/qemu/machine_unix.go:
```diff
@@ -33,6 +35,18 @@ func checkProcessStatus(processHint string, pid int, stderrBuf *bytes.Buffer) er
+}
+
+func useFdVLan() bool {
+	return strings.ToUpper(os.Getenv("CONTAINERS_USE_SOCKET_VLAN")) != "TRUE"
```
No objection to having an env flag for experimentation/debugging for now. However, it probably makes sense to drop the fd approach in the long term. Perhaps an intermediate step in a future PR would be to detect qemu < 7.2.0 and leave fd as a fallback. Either way, the more conservative approach this PR takes is good for now.

arixmkii replied:
I created a feature request for this functionality https://github.com/containers/podman/issues/18314

n1hility reviewed (changes requested):
Sorry for the delay, and never regret for providing thorough design / impl notes, these are very valuable, and helped us catch a bug in CI : ) 

This PR looks good, I left some minor suggestions. None are major blockers and can be addressed in follow-ups potentially

The only aspect that is necessary is a rebase, and you can drop the commit I added its in main now.

arixmkii commented on pkg/machine/qemu/machine.go:
```diff
@@ -507,29 +516,87 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
+	if isFdVlanVM {
+		// If the qemusocketpath exists and the vm is off/down, we should rm
+		// it before the dial as to avoid a segv
+		if err := vlanSocket.Delete(); err != nil {
```
I believe it is not a newly introduced race as it comes from the original code. I don't see how I can fix it, as this part is still pretty magical to me (was even more magical, when it reused QMP socket, still have memories from the days I started experimenting with this code). If in future this code will be phased out by Unix Socket VLAN, this would solve the race as well.

arixmkii commented on pkg/machine/qemu/machine.go:
```diff
@@ -1282,10 +1350,11 @@ func (v *MachineVM) setupAPIForwarding(cmd []string) ([]string, string, apiForwa
+	if err != nil {
+		return cmd, "", noForwarding
+	}
+	cmd = forwardSocketArgs(cmdWithPipe, socket.GetPath(), destSock, v.IdentityPath, forwardUser)
```
Moved the Pipe forwarding related call inside this function. Now the name is not as technically correct, but the code is easier to follow.

arixmkii commented on pkg/machine/qemu/machine.go:
```diff
@@ -507,29 +516,87 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
+
+	// Show migration recommendation if machine was created with different settings
+	if isFdVlanVM != useFdVLan() {
+		logrus.Warn("Stored Podman Machine configuration doesn't match current settings")
```
It checks that current setting in use is different from the time, when the machine was created. I updated the description (creating separate ones for either branch) and also demoted logging to DEBUG level to reduce noise (as both implementations should have same behavior on platforms, which supports both of them and no need to bother the user unless the troubleshooting happening).

These logs lines are pretty optional in general and can be freely removed. They raise from good intentions, but looks sort of trouble makers on their own, because tinkering with QEMU cmd in the machine config is sort of a feature of Podman.

n1hility replied:
ok so re-examining the overall logic, it seems the design CONTAINERS_USE_SOCKET_VLAN is very much a machine init (newmachine) feature, and really has no role on the start() path. So IMO its probably better to simplify the code path and just remove the extra logic here.

rhatdan commented:
@n1hility @arixmkii @baude @Luap99 @mheon What is going on with this one?

rhatdan commented:
@ashley-cui PTAL

arixmkii commented:
I addressed comments (except this logging, which could be helpful for migration cases, but it for sure don't have to be a part of this changeset at all costs and could be removed). I would appreciate some definitive list what to change before the next review iteration.

arixmkii commented:
@n1hility could you check this one once again and comment if I should change something additionally.

n1hility commented:
> @n1hility could you check this one once again and comment if I should change something additionally.

Just added some more feedback. Otherwise LGTM. Sorry I didn't reply until just now, I was out of office when you pinged me and just catching up.

n1hility commented:
BTW after addressing that last aspect i mentioned, while I think it looks good @baude and/or @ashley-cui may have thoughts as well.

arixmkii commented:
Did a rebase (had to re-arrange a number of things, because there were code changes), but now I have some issues running it locally (probably related to something else (including a different test box config)), will push the rebased version, when they are resolved.

arixmkii commented:
Pushed updated version. Some things feel like all over the place (comparing to previous version) because now there is a logic split between machine.go and config.go (this was introduced by some of the refactorings), but I tried to make it reasonable.

What I tested locally:
* FD socket on MacOS arm64
* VLAN socket on MacOS arm64 (passing env var to init)
* VLAN socket on Windows amd64

arixmkii commented:
According to test runner logs - machine tests are running on Fedora 38 and according to this page https://src.fedoraproject.org/rpms/qemu Fedora 38 has QEMU 7.2.1 stable version. It should be possible to expand this test with a matrix to test both FD and Unix domain socket modes. But I would prefer to experiment with it in a separate PR and have some hints how to run only minimal subset of jobs, while prototyping (I guess I can just directly hack the cirrus.yml removing parts, but there should be a better option, I guess).

arixmkii commented:
Rebased to latest main. Had to adapt to the new layout (methods instead of functions) and do some refactoring providing Windows specific alternatives to Unix specific bits.

arixmkii commented:
Will fix linter issues tomorrow and update again.

arixmkii commented:
Rebased and updated for latest codebase changes

arixmkii commented:
There is an yet unresolved issue with named pipe forwarding on Windows (not affecting other platforms obviously), because of https://github.com/containers/gvisor-tap-vsock/issues/276

Red checks seems to be Java related according to error logs, probably some infra related issues.

arixmkii commented:
Added workaround for gvproxy command line processing.

arixmkii commented:
Added fix for flaky issues during machine stop sequence (how the process kill is called and file access to file owned by another process on Windows platform).

arixmkii commented:
Added config to execute e2e tests on macOS.

Ran both versions w/ fd vlan and w/ UNIX domain socket vlan (actual command line verified during run using `ps` tool).

All tests passed for both flavors on M1 Pro MBP machine.
```
Ran 34 of 36 Specs in 813.328 seconds
SUCCESS! -- 34 Passed | 0 Failed | 0 Pending | 2 Skipped
PASS
```

rhatdan commented:
@n1hility PTAL

n1hility commented:
wow @arixmkii  you might be the most patient PR submitter ever. I promise to review this again this weekend

arixmkii commented:
@n1hility some context how the code evolved. This PR was updated to match the refactoring changes, this also lead to additional fixes, where non cross-platform changes were introduced. If it is desired I can factor these changes into a separate commit or PR for better traceability.

Some other things, which happened, while it was in review. OS version and QEMU versions (we are interested in QEMU > 7.2):
* Centos 8 stream - 6.2.0
* Centos 9 stream - 8.1.0
* Fedora 37 (soon to be EOL) - 7.0.0
* Fedora 38 - 7.2.6
* Fedora 39 - 8.1.1
* Debian Bullseye - 5.2 (7.2 available in backports)
* Debian Bookworm - 7.2
* Debian Trixie - 8.1.2
* MacOS brew - 8.1.2
* Windows - 8.1.2

So, it looks like having both FD and UNIX_VLAN could be only necessary for Centos 8 stream/RHEL 8 and it is not clear if it will receive Podman feature releases going forward. May be you if this will target Podman 5 (next major), then it is worth removing conditional options and provide migration path instead.

n1hility reviewed (changes requested):
Mostly looks good to me. just a couple notes

arixmkii commented on pkg/machine/qemu/machine.go:
```diff
@@ -615,23 +616,57 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
 	}
-	defer qemuSocketConn.Close()
+	defer func() {
+		_ = vlanSocket.Delete()
```
I 100% agree here. I replicated the old behavior, where my understanding was the goal of reusing qmp socket path and then deleting is was to get some "mode security" by not allowing to connect to that socket after the connection is established. It's pretty wrong to delete the resource, which actually was allocated by another process. If the original design decision was made only by the fact that "we already have some path prepared - let's use it and just free it afterwards", then it is fine by me to change it to a less surprising implementation. Will fix this.

arixmkii replied:
I removed the part and it uncovered another bug with gvproxy being terminated with SIGKILL would keep the socket in the filesystem. So, I have to change it to SIGTERM. I also added a check to startup that the file is not present or is successfully cleared before starting gvproxy. For Windows it will still be WM_QUIT message, I'm yet to run tests on a Windows box if this works correctly, but this can be cleared in a follow up. Here is a diff for review:

```diff
diff --git a/pkg/machine/qemu/machine.go b/pkg/machine/qemu/machine.go
index b80e54509..49e46737e 100644
--- a/pkg/machine/qemu/machine.go
+++ b/pkg/machine/qemu/machine.go
@@ -536,6 +536,10 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
 	if err != nil {
 		return err
 	}
+	err = vlanSocket.Delete()
+	if err != nil {
+		return err
+	}
 	isFdVlanVM := false
 	for _, c := range v.CmdLine {
 		if c == fdVlanNetdev() {
@@ -547,9 +551,6 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
 	if err != nil {
 		return fmt.Errorf("unable to start host networking: %q", err)
 	}
-	defer func() {
-		_ = vlanSocket.Delete()
-	}()
 
 	if isFdVlanVM {
 		qemuSocketConn, err = machine.DialSocketWithBackoffs(maxBackoffs, defaultBackoff, vlanSocket.GetPath())
@@ -805,7 +806,7 @@ func (v *MachineVM) ProxyPID() (int, error) {
 // cleanupVMProxyProcess kills the proxy process
 func (v *MachineVM) cleanupVMProxyProcess(proxyPid int) error {
 	// Kill the process
-	if err := killProcess(proxyPid); err != nil {
+	if err := killProcess(proxyPid, false); err != nil {
 		return err
 	}
 	if err := v.PidFilePath.Delete(); err != nil {
@@ -853,7 +854,7 @@ func (v *MachineVM) Stop(_ string, _ machine.StopOptions) error {
 		return stopErr
 	}
 
-	if err := killProcess(qemuPid); err != nil {
+	if err := killProcess(qemuPid, true); err != nil {
 		if stopErr == nil {
 			return err
 		}
@@ -1220,7 +1221,7 @@ func (v *MachineVM) startHostNetworking(vlanSocket *define.VMFile) (string, mach
 	}
 
 	cmd := gvproxy.NewGvproxyCommand()
-	cmd.AddQemuSocket(fmt.Sprintf("unix://%s", strings.ReplaceAll(vlanSocket.GetPath(), "\\", "/")))
+	cmd.AddQemuSocket(fmt.Sprintf("unix://%s", filepath.ToSlash(vlanSocket.GetPath())))
 	cmd.PidFile = v.PidFilePath.GetPath()
 	cmd.SSHPort = v.Port
 
diff --git a/pkg/machine/qemu/machine_unix.go b/pkg/machine/qemu/machine_unix.go
index 385611117..ef47b94df 100644
--- a/pkg/machine/qemu/machine_unix.go
+++ b/pkg/machine/qemu/machine_unix.go
@@ -34,8 +34,12 @@ func pingProcess(pid int) (int, error) {
 	return pid, nil
 }
 
-func killProcess(pid int) error {
-	return unix.Kill(pid, unix.SIGKILL)
+func killProcess(pid int, force bool) error {
+	if force {
+		return unix.Kill(pid, unix.SIGKILL)
+	} else {
+		return unix.Kill(pid, unix.SIGTERM)
+	}
 }
 
 func checkProcessStatus(processHint string, pid int, stderrBuf *bytes.Buffer) error {
diff --git a/pkg/machine/qemu/machine_windows.go b/pkg/machine/qemu/machine_windows.go
index 9692b8eaa..8c5da41cb 100644
--- a/pkg/machine/qemu/machine_windows.go
+++ b/pkg/machine/qemu/machine_windows.go
@@ -26,7 +26,7 @@ func pingProcess(pid int) (int, error) {
 	return pid, nil
 }
 
-func killProcess(pid int) error {
+func killProcess(pid int, force bool) error {
 	machine.SendQuit(uint32(pid))
 	return nil
 }
```

n1hility replied:
That looks good! BTW I have a consolidated win quit lib that gvproxy is now using here:https://github.com/containers/winquit

See the client example which also has a call that will attempt QUIT then fall back to kill after a timeout.
It's not critical of this PR just a mention.

n1hility replied:
sure yeah not necessary just mentioning

arixmkii commented on pkg/machine/qemu/machine.go:
```diff
@@ -1240,18 +1293,18 @@ func getDiskSize(path string) (uint64, error) {
 
 	cmd := gvproxy.NewGvproxyCommand()
-	cmd.AddQemuSocket(fmt.Sprintf("unix://%s", v.QMPMonitor.Address.GetPath()))
+	cmd.AddQemuSocket(fmt.Sprintf("unix://%s", strings.ReplaceAll(vlanSocket.GetPath(), "\\", "/")))
```
Fixed.

rhatdan commented:
@baude @ashley-cui PTAL

baude commented on pkg/machine/qemu/command.go:
```diff
@@ -41,10 +41,29 @@ func (q *QemuCmd) SetQmpMonitor(monitor Monitor) {
 
 // SetNetwork adds a network device to the machine
-func (q *QemuCmd) SetNetwork() {
+func (q *QemuCmd) SetNetwork(name string) error {
```
i would prefer vlanName here ... soft ask.

baude commented on pkg/machine/qemu/command.go:
```diff
@@ -41,10 +41,29 @@ func (q *QemuCmd) SetQmpMonitor(monitor Monitor) {
+			*q = append(*q, []string{"-netdev", socketVlanNetdev(vlanSocket.GetPath())}...)
+		}
+	}
+	*q = append(*q, []string{"-device", "virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee"}...)
```
break MAC to const

baude commented on pkg/machine/qemu/command.go:
```diff
@@ -41,10 +41,29 @@ func (q *QemuCmd) SetQmpMonitor(monitor Monitor) {
+	return nil
+}
+
+func fdVlanNetdev() string {
```
const or var?

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -531,23 +532,58 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
+		return fmt.Errorf("unable to start host networking: %q", err)
+	}
+
+	if isFdVlanVM {
```
break this out to a new function ...

arixmkii replied:
This doesn't play well. We have to `defer` invocations in this block. The best the body would look after the change is something like:
```go
		fd, qemuSocketConn, err = getVlanFd(vlanSocket.GetPath())
		if qemuSocketConn != nil {
			defer qemuSocketConn.Close()
		}
		if err != nil {
			return nil
		}
		defer fd.Close()
```

It is not compact, neither it is a lot more readable than the current thing. Especially processing one of the output parameters in both error and success cases is looking confusing (at least for a non-golang dev). I can code change this if you insist. Will address the else branch as requested.

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -531,23 +532,58 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
+		}
+		defer fd.Close()
+	} else {
+		time.Sleep(defaultBackoff)
```
same here

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -557,7 +593,10 @@ func (v *MachineVM) Start(name string, opts machine.StartOptions) error {
 	attr := new(os.ProcAttr)
-	files := []*os.File{dnr, dnw, dnw, fd}
+	files := []*os.File{dnr, dnw, dnw}
+	if fd != nil {
```
comment on this check.

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -762,15 +803,14 @@ func (v *MachineVM) ProxyPID() (int, error) {
-// cleanupVMProxyProcess kills the proxy process and removes the VM's pidfile
-func (v *MachineVM) cleanupVMProxyProcess(proxyProc *os.Process) error {
+// cleanupVMProxyProcess kills the proxy process
+func (v *MachineVM) cleanupVMProxyProcess(proxyPid int) error {
```
out of curiosity, why did you choose to change the type here from something reasonably strong to something more generic like an int?  im not sure i have an objection with it, i would however like to know.

arixmkii replied:
Because internally this `os.Process` pointer was used only to call `Kill` method https://pkg.go.dev/os#Process.Kill, which really doesn't work as we need to send SIGTERM for correct cleanup and it also doesn't work on Windows. There is no point extracting the pid here from `os.Proccess` as previously we acquired that looking up pid value we already stored as part of the state. `os.Process` really brings no benefits there if we can't use its API.

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -921,7 +961,11 @@ func (v *MachineVM) stopLocked() error {
 
 	fmt.Println("Waiting for VM to exit...")
-	for isProcessAlive(vmPid) {
+	for {
```
is it possible that we never exit this loop

arixmkii replied:
Added retry limit

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -930,32 +974,42 @@ func (v *MachineVM) stopLocked() error {
-		Timeout: timeout,
-	}
-	return monitor, nil
+	return define.NewMachineFile(filepath.Join(rtDir, name+".sock"), nil)
```
fyi, there are actually plans for podman 5 to make a sister to `VMFile` but more like VMSocket or MachineSocket.

arixmkii replied:
I will refactor this PR if the changes lands before this one is merged.

baude commented on pkg/machine/qemu/machine.go:
```diff
@@ -1156,18 +1210,18 @@ func getDiskSize(path string) (uint64, error) {
 
 	cmd := gvproxy.NewGvproxyCommand()
-	cmd.AddQemuSocket(fmt.Sprintf("unix://%s", v.QMPMonitor.Address.GetPath()))
+	cmd.AddQemuSocket(fmt.Sprintf("unix://%s", filepath.ToSlash(vlanSocket.GetPath())))
```
nice!

baude commented:
I would love, where possible, to see more tests that lean heavily on your code.  Unittests where possible as well.  This is a lot of code to change and it is going to get heavily reworked and broken in the very near future.  Would you consider going back and accessing your comfort with tests and that regressions are not introduced?

baude commented:
And finally, I don't recall seeing any sort of version checking but yet this requires a newer QEMU if I understood things correctly.  Podman, Podman Desktop, Brew, Distros are really struggling with package level interdependencies.  Do you feel that we should somehow do something in your code to avoid this?

baude commented:
ps.  i think we just about have hyperv and wsl passing tests again (maybe a flake here and there) ... i would prefer that we do not merge this PR until #20953 is merged.

arixmkii commented:
> Would you consider going back and accessing your comfort with tests and that regressions are not introduced?

Yes, I will try to add tests after addressing your other comments.

> ps. i think we just about have hyperv and wsl passing tests again (maybe a flake here and there)

I'm counting on it. It was just a coincidence that I needed a rebase (because line endings were changes in README.md from CRLF to LF).

> And finally, I don't recall seeing any sort of version checking but yet this requires a newer QEMU if I understood things correctly.

That newer version is one, which was released a year ago :) I have a version specific breakdown in this comment https://github.com/containers/podman/pull/17473#issuecomment-1817524290 Also, this code path is enabled unconditionally only for Windows code and is behind Env var feature toggle for others. I will think about if it is viable to add version check within current changes. Will either add them or create an issue for a follow up and assign to me.

arixmkii commented:
Addressed comments. Will try to unit add tests later this week.

arixmkii commented:
@baude I rebased to latest changes and added unit tests, which are similar to existent ones. I see that currently there are not many unit tests and machines are mostly tested e2e.

I wanted to update code to use your new utilities from `sockets`, but I have question. Why is the new ready socket different from qmp socket. The QMP one is using different runtime directory depending on rootless or rootful configurations https://github.com/containers/podman/blob/c324dbb8b4bf094aa6b222bf60ee6fcf5e2f96aa/pkg/machine/qemu/machine.go#L913 but the new Ready socket is always using temp directory w/o any conditional checks on the isRootful state https://github.com/containers/podman/blob/c324dbb8b4bf094aa6b222bf60ee6fcf5e2f96aa/pkg/machine/qemu/config.go#L137 Could you clarify, which one I would use for vlan sort of socket? Or is it a bug introduced with the refactoring to ReadySocket?

arixmkii commented:
Will be continued in https://github.com/containers/podman/pull/21594
